File: python/search/text_match.py
from pymilvus import MilvusClient, DataType, Function, FunctionType


client = MilvusClient(
    uri="http://localhost:19530",
    token="root:Milvus"
)

def create_schema():
    schema = MilvusClient.create_schema(enable_dynamic_field=False)
    schema.add_field(
        field_name="id",
        datatype=DataType.INT64,
        is_primary=True,
        auto_id=True
    )
    analyzer_params = {"type": "english"}
    schema.add_field(
        field_name='text',
        datatype=DataType.VARCHAR,
        max_length=1000,
        enable_analyzer=True,  # Whether to enable text analysis for this field
        analyzer_params=analyzer_params,
        enable_match=True  # Whether to enable text match
    )
    schema.add_field(
        field_name="embeddings",
        datatype=DataType.FLOAT_VECTOR,
        dim=5
    )
    return schema

def create_index():
    index_params = client.prepare_index_params()

    index_params.add_index(
        field_name="embeddings",
        index_type="AUTOINDEX",
        metric_type="L2"
    )
    return index_params

def create_collection(schema, index_params):
    client.drop_collection(collection_name="text_match_collection")
    client.create_collection(
        collection_name='text_match_collection',
        schema=schema,
        index_params=index_params
    )

def insert():
    client.insert('text_match_collection', [
        {'text': 'this is keyword1.', "embeddings": [0.1, 0.2, 0.3, 0.4, 0.5]},
        {'text': 'no keyword.', "embeddings": [0.2, 0.3, 0.4, 0.5, 0.6]},
        {'text': 'this keyword1 and keyword2.', "embeddings": [0.3, 0.4, 0.5, 0.6, 0.7]},
    ])
    client.flush(collection_name="text_match_collection")
    res = client.query(collection_name="text_match_collection",
                       filter="",
                       output_fields=["count(*)"],
                       consistency_level="Strong")
    print(res)

def search(filter: str):
    res = client.search(
        collection_name='text_match_collection',
        data=[[0.1, 0.2, 0.3, 0.5, 0.7]],
        anns_field='embeddings',
        filter=filter,
        limit=3,
        search_params={"params": {"nprobe": 10}},
        output_fields=["id", "text"],
    )
    for hits in res:
        print("TopK results:")
        for hit in hits:
            print(hit)

    result = client.query(
        collection_name="text_match_collection",
        filter=filter,
        output_fields=["id", "text"]
    )
    print(result)

if __name__ == "__main__":
    schema = create_schema()
    index_params = create_index()
    create_collection(schema, index_params)
    insert()

    search("TEXT_MATCH(text, 'keyword1 keyword2')")
    search("TEXT_MATCH(text, 'keyword1') and TEXT_MATCH(text, 'keyword2')")

File: go/search/text_match.go
package search

import (
	"context"
	"fmt"

	"github.com/milvus-go-examples/util"
	"github.com/milvus-io/milvus/client/v2/entity"
	"github.com/milvus-io/milvus/client/v2/index"
	"github.com/milvus-io/milvus/client/v2/milvusclient"
)

func TextMatch() {
	createCollectionText()
	defer util.DropCollection("my_collection")

	textMatchSearch()
}

func createCollectionText() {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	client, err := util.GetClient(ctx)
	if err != nil {
		fmt.Println(err.Error())
	}
	defer client.Close(ctx)

	analyzerParams := map[string]any{"type": "english"}

	schema := entity.NewSchema().WithDynamicFieldEnabled(false)
	schema.WithField(entity.NewField().
		WithName("id").
		WithDataType(entity.FieldTypeInt64).
		WithIsPrimaryKey(true).
		WithIsAutoID(true),
	).WithField(entity.NewField().
		WithName("text").
		WithDataType(entity.FieldTypeVarChar).
		WithEnableAnalyzer(true).
		WithEnableMatch(true).
		WithAnalyzerParams(analyzerParams).
		WithMaxLength(1000),
	).WithField(entity.NewField().
		WithName("embeddings").
		WithDataType(entity.FieldTypeFloatVector).
		WithDim(5),
	)

	indexOption := milvusclient.NewCreateIndexOption("my_collection", "embeddings",
		index.NewAutoIndex(entity.MetricType(entity.L2)))

	err = client.CreateCollection(ctx,
		milvusclient.NewCreateCollectionOption("my_collection", schema).
			WithIndexOptions(indexOption))
	if err != nil {
		fmt.Println(err.Error())
		// handle error
	}

	_, err = client.Insert(ctx, milvusclient.NewColumnBasedInsertOption("my_collection").
		WithVarcharColumn("text", []string{
			"this is keyword1.",
			"no keyword.",
			"keyword1 and keyword2.",
		}).
		WithFloatVectorColumn("embeddings", 5, [][]float32{
			{0.1, 0.2, 0.3, 0.4, 0.5},
			{0.2, 0.3, 0.4, 0.5, 0.6},
			{0.3, 0.4, 0.5, 0.6, 0.7},
		}))
	if err != nil {
		fmt.Println(err.Error())
		// handle err
	}

	util.FlushLoadCollection(client, "my_collection")
}

func textMatchSearch() {
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	client, err := util.GetClient(ctx)
	if err != nil {
		fmt.Println(err.Error())
	}
	defer client.Close(ctx)

	queryVector := []float32{0.1, 0.2, 0.3, 0.5, 0.7}
	filter := "TEXT_MATCH(text, 'keyword1 keyword2')"

	resultSets, err := client.Search(ctx, milvusclient.NewSearchOption(
		"my_collection", // collectionName
		10,              // limit
		[]entity.Vector{entity.FloatVector(queryVector)},
	).WithANNSField("embeddings").
		WithFilter(filter).
		WithOutputFields("id", "text"))
	if err != nil {
		fmt.Println(err.Error())
		// handle error
	}

	for _, resultSet := range resultSets {
		fmt.Println("IDs: ", resultSet.IDs.FieldData().GetScalars())
		fmt.Println("Scores: ", resultSet.Scores)
		fmt.Println("text: ", resultSet.GetColumn("text").FieldData().GetScalars())
	}

	filter = "TEXT_MATCH(text, 'keyword1') and TEXT_MATCH(text, 'keyword2')"
	resultSet, err := client.Query(ctx, milvusclient.NewQueryOption("my_collection").
		WithFilter(filter).
		WithOutputFields("id", "text"))
	if err != nil {
		fmt.Println(err.Error())
		// handle error
	}

	fmt.Println("id: ", resultSet.GetColumn("id").FieldData().GetScalars())
	fmt.Println("text: ", resultSet.GetColumn("text").FieldData().GetScalars())
}
